lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,427 @@
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
+
+    """
+
+    model: QuantumModel
+
+    def __init__(
+        self,
+        model: QuantumModel,
+        hyperparams: dict[str, float],
+        loss_fn: Callable[[Any, Any], float],
+        bounds: ArrayLike | None = None,
+    ) -> None:
+        """Initialise the Nelder-Mead optimizer.
+
+        The hyperparameters may contain the following key value pairs:
+
+        - `adaptive`: Adjust the algorithm's parameters based on the
+                dimensionality of the problem. This adaptation is
+                particularly helpful when minimizing functions in
+                high-dimensional spaces, bool.
+
+        - `maxfev`: Maximum number of function evaluations allowed.
+                Default is 1000. int.
+
+        - `initial_simplex`: If provided, the `initial simplex`
+                replaces the initial model weights. Each row
+                initial_simplex[i, :] should contain the coordinates
+                of the ith vertex among the N+1 vertices in the
+                simplex, where N represents the dimension,
+                ArrayLike (N+1, N), float.
+
+        - `xatol`: The acceptable level of absolute error in the
+                optimal model weights (optimal solution) between
+                iterations that indicates convergence, float.
+
+        - `fatol`: The acceptable level of absolute error in the
+                loss value between iterations that indicates convergence,
+                float.
+        }
+
+        Parameters
+        ----------
+        model : :py:class:`.QuantumModel`
+            A lambeq quantum model.
+        hyperparams : dict of str to float.
+            A dictionary containing the models hyperparameters.
+        loss_fn : Callable
+            A loss function of form `loss(prediction, labels)`.
+        bounds : ArrayLike, optional
+            The range of each of the model parameters.
+
+        Raises
+        ------
+        ValueError
+            - If the hyperparameters are not set correctly, or if the
+            length of `bounds` does not match the number of the model
+            parameters.
+            - If the lower bounds are greater than the upper bounds.
+            - If the initial simplex is not a 2D array.
+            - If the initial simplex does not have N+1 rows, where N is
+            the number of model parameters.
+
+        Warning
+            - If the initial model weights are not within the bounds.
+
+        References
+        ----------
+        Gao, Fuchang & Han, Lixing. (2012). Implementing the Nelder-Mead
+        simplex algorithm with adaptive parameters.
+        Computational Optimization and Applications. 51. 259-277.
+        10.1007/s10589-010-9329-3.
+
+        """
+        super().__init__(model, hyperparams, loss_fn, bounds)
+
+        self.ncalls = 0
+
+        def objective(x, y, w):
```

Comment:
Is it possible to add type annotations here?

Reply:
It will be complex I guess. When I tried with `x: ArrayLike`, mypy resulted in the following error:
```
Argument 1 has incompatible type "Iterable[Any]"; expected
"Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"  [arg-type]
```